Bluetooth: Controller: Add NRF_CCM support in nRF54L15 SoC

cvinayak opened the pull request:
Add NRF_CCM h/w based encryption support in nRF54L15 SoCs.

cvinayak committed:
Bluetooth: Controller: Single timer use Extended Advertising nRF54L support

Update Extended Advertising for single timer use support
required for nRF54L Series SoCs.

As the timer is reset on every radio end, add implementation
to accumulate the last_pdu_end_us.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Add NRF_CCM support in nRF54L15 SoC

Add NRF_CCM h/w based encryption support in nRF54L15 SoCs.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Minor fix defines and conditional compilations

Replace use of HAL_RADIO_RECV_TIMEOUT_CANCEL_PPI with
HAL_TRIGGER_CRYPT_PPI; they have same value but use the
correctly named HAL_TRIGGER_CRYPT_PPI define now.

Fix conditional compile to correctly exclude reference of
NRF_CCM when CONFIG_BT_CTLR_PRIVACY=n.

Fix missing reference to CCM_TASKS_CRYPT_DELAY_BITS when
CONFIG_BT_CTLR_SW_SWITCH_SINGLE_TIMER=y.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Align DDPI usage with nRF54L power domains

Align DDPI usage with nRF54L series power domains and
available DDPI configurations.

Use indices below 4 when needing to bridge with GPIO,
use indices below 8 when needing to bridge with MCU domain
like CCM and AAR; and use indices above 7 to bridge with
Peripheral domain.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Fix radio_tmr_start_us for single timer use

This commit addresses two bugs in use of single timer s/w
switch implementation, incorrect aux_offset in subsequent
ADV_EXT_IND and ISO receiver failing to receive second or
greater BIS subevents.

Fix radio_tmr_start_now implementation to consider the
initial latency in starting the event timer in single timer
use mode. This fixes the incorrect aux_offset in ADV_EXT_IND
PDUs.

Fix radio_tmr_start_us implementation for single timer use,
as the timer is reset on every radio end, and hence the
requested start_us has to be adjusted for the accumulated
last_end_pdu_us value. This fixes the BIS subevent
receptions.

Also, fix the maximum radio ISR latency value used in
radio_tmr_start_us to consider the maximum Rx chain delay
and maximum radio ramp up delays. 80 us + ~30 us + ~40 us
should be able to meet the 150 us tIFS duration.

Relates to commit bcd28e0a866b ("Bluetooth: Controller: Fix
sw switch single timer for spurious TXEN/RXEN").

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
tests: bsim: Bluetooth: nrRF53/54L: Throughput and multiple_id testing

Enable throughput and multiple_id testing on nRF5340bsim and
nrf54l15bsim.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: BUILD_ASSERT ticker nodes exceed UINT8_MAX

Added BUILD_ASSERT to ensure ticker node count does not
exceed UINT8_MAX.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Single timer use ISO Peripheral nRF54L support

Update ISO Peripheral for single timer use support required
for nRF54L Series SoCs.

As the timer is reset on every radio end, use the
radio_tmr_hcto_configure_abs() function to set the header
complete timeout correctly when using single timer for s/w
switching.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Single timer use ISO Receive nRF54L support

Update ISO Sync Receiver for single timer use support
required for nRF54L Series SoCs.

As the timer is reset on every radio end, use the
radio_tmr_hcto_configure_abs() function to set the header
complete timeout correctly when using single timer for s/w
switching.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Single timer use ISO Central nRF54L support

Update ISO Central for single timer use support required
for nRF54L Series SoCs.

As the timer is reset on every radio end, use the
radio_tmr_hcto_configure_abs() function to set the header
complete timeout correctly when using single timer for s/w
switching.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Single timer use Extended Scanning nRF54L support

Update Extended Scanning for single timer use support
required for nRF54L Series SoCs.

As the timer is reset on every radio end, use the
radio_tmr_hcto_configure_abs() function to set the header
complete timeout correctly when using single timer for s/w
switching.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Single timer use Conn Scheduling nRF54L support

Update advanced non-overlapping Central Connection Creation
scheduling for single timer use support required for nRF54L
Series SoCs.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Single timer use Periodic Adv Sync nRF54L support

Update Periodic Advertising Sync implementation for single
timer use support required for nRF54L Series SoCs.

As the timer is reset on every radio end, use the
radio_tmr_hcto_configure_abs() function to set the header
complete timeout correctly when using single timer for s/w
switching.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Fix single timer s/w switch when using DPPI

Fix single timer s/w switch when using DPPI on nRF53/54L
where the previous compare register index is incorrect when
the base index is a odd numbered DPPI used.

The issue is discovered as failing ISO Receiver sample with
Coded PHY and single timer use, where the compare base is
configured to use the DPPI 3 (instead of 4).

Upstream samples do not use single timer for s/w switch, the
issue is discovered when adding support for nRF54L which has
to use single timer due to the radio only having a single
timer.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Fix nRF5 radio and event timer abstractions

Fix nRF5 radio and event timer abstractions for some that
where hard coded compared to use of the ones defined in the
resources header file.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Fix single timer s/w switch base indices

Fix single timer s/w switch base indices from overlapping
with sample index.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Revert "Bluetooth: Controller: Fix ext conn create when using single timer"

This reverts commit a7fff7445cd34d8388c13fcee430635252bcb044.

The next commit will re-add back the fix to primary
advertising channel plus auxiliary advertising channel too.
This will help keep the fix together regarding use of
single timer for software tIFS switching implementation.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Restrict ISO Sync Receiver subevent jitter

Restrict the jitter considered in the PDU reception in the
subevents of ISO Sync Receiver.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

aescolar commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio.c:
```diff
@@ -641,15 +641,6 @@ uint32_t radio_is_done(void)
 }
 #endif /* !CONFIG_BT_CTLR_SW_SWITCH_SINGLE_TIMER */
 
-uint32_t radio_is_tx_done(void)
```
It would be nice if this revert commit elaborated on why it is doing the revert, specially as the next commit reverts the revert :)

aescolar commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_sync.c:
```diff
@@ -757,7 +757,7 @@ static void isr_aux_setup(void *param)
 	hcto += radio_rx_chain_delay_get(phy_aux, PHY_FLAGS_S8);
 	hcto += addr_us_get(phy_aux);
-	radio_tmr_hcto_configure(hcto);
+	radio_tmr_hcto_configure_abs(hcto);
```
Silly comment: quite a few very similar 1 line commits :)

aescolar reviewed (commented):
Thanks @cvinayak, (a relatively quick review given the PR size)
one relevant comment (the one about removing the sim hal files) and otherwise just some suggestions

aescolar commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio_sim_nrf52.h:
I see.
But this change is making me very uneasy.
* we are testing in CI only each target against itself, so it can be easy to mask an error. (AFAIR Demant was testing its target against the nrf52 also, so their internal CI would be likely to detect an issue here @wopu-ot @thoh-ot could you guys spin this PR thru your tests?)
* In this PR we are changing quite a lot of timings, so it is not trivial to see if something went a bit/a lot bad due to the hal file merge. It would be much nicer to decouple a change like this into another PR (also in case we need to revert)
* The majority of timing differences are sub µs. But some of them are >5µs. That includes the S2 coded with differences of 10µs in the rampup time and 5µs in the Rx chain delay. But I tend to recall we are not testing S2 in CI(?)
I would sleep much better if we did not do this merge in this PR.

aescolar replied:
> simulation specific incorrect timings with respect to actual target values is removed.

The timing constants in the hal files for the simulation targets are based on the actual timings of the HW models which are in turn mostly based on what the spec said at the time. These hal timings are not incorrect as they reflect the timing of the models.
While, I understand, the timings for the real target hal files were based on measurements of those targets.
Meaning, those hal files timing differences actually reflect different targets' timings. Now we could argue the models should match those measurements from real target instead of the whatever the spec said, but that is another discussion that would require changing the models timings and not just these hal files.

> The removed incorrect values in simulation specific file are contained in a single commit: https://github.com/zephyrproject-rtos/zephyr/commit/063b50a75ec8c194a818eb8de31ae03d8a68afa8

Yes, but the very next commit (89f32146083cf508b898287e94b3fb063220a40c) changes what is expected from the hal files, so it is not like that commit can be just reverted without doing quite a few other changes.

Could you please not do this change in this PR?

cvinayak replied:
All this is skeptical that bsim testing is green (https://github.com/zephyrproject-rtos/zephyr/actions/runs/12604177426/job/35130591570?pr=83394) with the real target hal files, makes me wonder if the results are all false positive, or the timings constant just the similar values but in redundant simulation hal files... hmmm!

Ok, i can revert back and add the files back for your satisfaction, but I will need a new nrf54L15 simulation hal files so that at least the file structure/includes are consistent (today, there is no simulation hal for nRF54L upstream).

cvinayak committed:
Bluetooth: Controller: Add radio_sim_nrf54l.h file

Add radio_sim_nrf54l.h file, which for now a copy of the
SoC radio_nrf54lx.h file. This is added for having separate
simulation model dependent timing values and specific
implementation details.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
tests: bsim: Bluetooth: Workaround overlapping ACL in BAP broadcast sink

Workaround by using `-start_offset` to ensure that the ACL
established does not overlap the sink being established.

The behavior between nrf52_bsim and nrf54l15bsim is
different where in the two device are starting to
advertising at the same time in nrf54l15bsim where as
there is an offset already in nrf52_bsim test.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

aescolar commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio_sim_nrf54l.h:
This would not be required in this PR as this was already the case, but this timings are off for simulation as the simulated radio has the exact same timings for all SOCs so far,
If you want you can pick this commit already into this PR to correct the values
https://github.com/aescolar/zephyr/commit/ed78e93055dd603d3398ef6ff09441b8aa5ff627
Otherwise I can send a follow up PR with it.

cvinayak replied:
Follow up PR is good. Should your  RNG PR go in first or this one? I am ok with either.

aescolar replied:
Im fine either way. You are welcome to +1 the RNG PR if you think it is good :)

aescolar reviewed (approved):
+1
